Issue with Response OutputStream#close() rethrowing same EofException instance

**Jetty version(s)**

Jetty 12.0.7

**Jetty Environment**

core

**Java version/vendor** `(use: java -version)`

java version "21.0.1" 2023-10-17 LTS
Java(TM) SE Runtime Environment (build 21.0.1+12-LTS-29)
Java HotSpot(TM) 64-Bit Server VM (build 21.0.1+12-LTS-29, mixed mode, sharing)

**OS type/version**

ProductName:		macOS
ProductVersion:		14.4.1
BuildVersion:		23E224

**Description**

In the case where a client hangs up *before* a Jetty Handler has written a response, an `EofException` is raised when attempting to write to the response output stream.

This is expected.

It is best practice to acquire and release the response output stream using with the try-with-resources idiom:

```java
                   public boolean handle(Request request, Response response, Callback callback) {
                        try (final var body = Content.Sink.asOutputStream(response)) {
                            body.write("Hello World".getBytes(StandardCharsets.UTF_8));
                            callback.succeeded();
                        } catch (IOException e) {
                            callback.failed(e);
                        }
                        return true;
                    }
```

Note that the try with resources block is translated into a regular try block per the [Java Language Specification](https://docs.oracle.com/javase/specs/jls/se21/html/jls-14.html#jls-14.20.3.1), so when the block completes the `OutputStream#close()` method will be invoked.

The generated try block has the following form: 

```java
{
    final {VariableModifierNoFinal} R Identifier = Expression;
    Throwable #primaryExc = null;

    try ResourceSpecification_tail
        Block
    catch (Throwable #t) {
        #primaryExc = #t;
        throw #t;
    } finally {
        if (Identifier != null) {
            if (#primaryExc != null) {
                try {
                    Identifier.close();
                } catch (Throwable #suppressedExc) {
                    #primaryExc.addSuppressed(#suppressedExc);
                }
            } else {
                Identifier.close();
            }
        }
    }
}
```
- Note the line: `#primaryExc.addSuppressed(#suppressedExc);`

The logic of `Throwable#addSuppressed` is:

```java
public final synchronized void addSuppressed(Throwable exception) {
        if (exception == this)
            throw new IllegalArgumentException(SELF_SUPPRESSION_MESSAGE, exception);

        Objects.requireNonNull(exception, NULL_CAUSE_MESSAGE);

        if (suppressedExceptions == null) // Suppressed exceptions not recorded
            return;

        if (suppressedExceptions == SUPPRESSED_SENTINEL)
            suppressedExceptions = new ArrayList<>(1);

        suppressedExceptions.add(exception);
    }
```
- If the suppressed exception is the same instance as `this` then raise an `IlegalArgumentException` (to prevent an exception referring to itself as suppressed)

In Jetty 12 (to the best of my recollection this was not the case in earlier Jetty versions, but have not confirmed this), the `OutputStream#close()` call rethrows the exact same instance of `EofException` as was raised when the call to `OutputStream#write()` was invoked.

This causes the generated try block to end up raising an `IllegalArgumentException` due to `Throwable#addSuppressed()` refusing to add itself as a suppressed exception. 

Whether intentional or not, the generated try block excludes the possibility that the `close()` method may throw the exact same exception instance which originally caused the try with resources block to be exited.

I cannot think of a good reason for the `close()` method to rethrow the same exception instance. I think a failure in the `close()` method is semantically different to the failure in the `write()` method and should result in a different exception instance (if any) being thrown. 

In this specific case, where the client has hung-up,  I think the appropriate behaviour of `close()` is to complete without raising any error.

**How to reproduce?**

See the test case below:

```java
package com.example.jetty;

import org.eclipse.jetty.io.Content;
import org.eclipse.jetty.server.Handler;
import org.eclipse.jetty.server.Request;
import org.eclipse.jetty.server.Response;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.ServerConnector;
import org.eclipse.jetty.server.handler.ContextHandler;
import org.eclipse.jetty.util.Callback;

import java.io.IOException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

public class Main {

    private static void sleep(final int seconds) {
        try {
            TimeUnit.SECONDS.sleep(seconds);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public static void main(String... args) throws Exception {
        final var server = new Server();
        final var connector = new ServerConnector(server, 1, 1);
        server.addConnector(connector);
        final var context = new ContextHandler("/");

        final var clientResponse = new AtomicReference<CompletableFuture<HttpResponse<String>>>();

        context.setHandler(
                new Handler.Abstract() {

                    @Override
                    public boolean handle(Request request, Response response, Callback callback) {
                        try (final var body = Content.Sink.asOutputStream(response)) {
                            // simulate the client hanging up
                            if (clientResponse.get() != null) {
                                clientResponse.get().cancel(true);
                            }
                            sleep(2);

                            // Raises EofException because client hung up
                            body.write("Hello World".getBytes(StandardCharsets.UTF_8));
                            callback.succeeded();
                        } catch (IOException e) {
                            // Does NOT reach here, because the OutputStream#close()
                            // rethrows same EofException instance, which causes
                            // try with resources block call to: Throwable#addSuppresed() to
                            // fail with java.lang.IllegalArgumentException: Self-suppression not
                            // permitted
                            callback.failed(e);
                        }
                        return true;
                    }
                });
        server.setHandler(context);

        server.start();

        clientResponse.set(
                HttpClient.newHttpClient()
                        .sendAsync(
                                HttpRequest.newBuilder().GET().uri(server.getURI()).build(),
                                HttpResponse.BodyHandlers.ofString()));
        sleep(2);
        server.stop();
    }
}

```

This is the output of running the above test case:

```
[main] INFO org.eclipse.jetty.server.Server - jetty-12.0.7; built: 2024-02-29T21:19:41.771Z; git: c89aca8fd34083befd79f328a3b8b6ffff04347e; jvm 21.0.1+12-LTS-29
[main] INFO org.eclipse.jetty.server.handler.ContextHandler - Started oejsh.ContextHandler@cb644e{ROOT,/,b=null,a=AVAILABLE,h=cej.Main$@56ef9176{STARTED}}
[main] INFO org.eclipse.jetty.server.AbstractConnector - Started ServerConnector@56235b8e{HTTP/1.1, (http/1.1)}{0.0.0.0:57211}
[main] INFO org.eclipse.jetty.server.Server - Started oejs.Server@7921b0a2{STARTING}[12.0.7,sto=0] @186ms
[main] INFO org.eclipse.jetty.server.Server - Stopped oejs.Server@7921b0a2{STOPPING}[12.0.7,sto=0]
[main] INFO org.eclipse.jetty.server.AbstractConnector - Stopped ServerConnector@56235b8e{HTTP/1.1, (http/1.1)}{0.0.0.0:0}
[qtp1282788025-25] WARN org.eclipse.jetty.server.Response - writeError: status=500, message=java.lang.IllegalArgumentException: Self-suppression not permitted, response=org.eclipse.jetty.server.handler.ContextResponse@7ad5cbae
java.lang.IllegalArgumentException: Self-suppression not permitted
	at java.base/java.lang.Throwable.addSuppressed(Throwable.java:1096)
	at com.example.jetty.rethrown.exception.issue/com.example.jetty.Main$1.handle(Main.java:44)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.Server.handle(Server.java:179)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:619)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:411)
	at org.eclipse.jetty.io@12.0.7/org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io@12.0.7/org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io@12.0.7/org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util@12.0.7/org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:971)
	at org.eclipse.jetty.util@12.0.7/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1201)
	at org.eclipse.jetty.util@12.0.7/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1156)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.eclipse.jetty.io.EofException
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpConnection$SendCallback.reset(HttpConnection.java:739)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpConnection$HttpStreamOverHTTP1.send(HttpConnection.java:1418)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpChannelState$ChannelResponse.write(HttpChannelState.java:1246)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.Response$Wrapper.write(Response.java:751)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.handler.ContextResponse.write(ContextResponse.java:56)
	at org.eclipse.jetty.io@12.0.7/org.eclipse.jetty.io.content.ContentSinkOutputStream.write(ContentSinkOutputStream.java:53)
	at java.base/java.io.OutputStream.write(OutputStream.java:124)
	at com.example.jetty.rethrown.exception.issue/com.example.jetty.Main$1.handle(Main.java:52)
	... 11 more
[qtp1282788025-25] WARN org.eclipse.jetty.server.Response - writeError: status=500, message=java.lang.IllegalArgumentException: Self-suppression not permitted, response=ErrorResponse@7481e31b{500,GET@0 http://10.20.1.40:57211/ HTTP/1.1}
java.lang.IllegalArgumentException: Self-suppression not permitted
	at java.base/java.lang.Throwable.addSuppressed(Throwable.java:1096)
	at com.example.jetty.rethrown.exception.issue/com.example.jetty.Main$1.handle(Main.java:44)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.Server.handle(Server.java:179)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:619)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:411)
	at org.eclipse.jetty.io@12.0.7/org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io@12.0.7/org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io@12.0.7/org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util@12.0.7/org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:971)
	at org.eclipse.jetty.util@12.0.7/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1201)
	at org.eclipse.jetty.util@12.0.7/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1156)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.eclipse.jetty.io.EofException
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpConnection$SendCallback.reset(HttpConnection.java:739)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpConnection$HttpStreamOverHTTP1.send(HttpConnection.java:1418)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.internal.HttpChannelState$ChannelResponse.write(HttpChannelState.java:1246)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.Response$Wrapper.write(Response.java:751)
	at org.eclipse.jetty.server@12.0.7/org.eclipse.jetty.server.handler.ContextResponse.write(ContextResponse.java:56)
	at org.eclipse.jetty.io@12.0.7/org.eclipse.jetty.io.content.ContentSinkOutputStream.write(ContentSinkOutputStream.java:53)
	at java.base/java.io.OutputStream.write(OutputStream.java:124)
	at com.example.jetty.rethrown.exception.issue/com.example.jetty.Main$1.handle(Main.java:52)
	... 11 more

```

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkOutputStream.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkOutputStream.java
@@ -33,6 +33,7 @@ public class ContentSinkOutputStream extends OutputStream
 {
     private final Blocker.Shared _blocking = new Blocker.Shared();
     private final Content.Sink sink;
+    private boolean failed;
 
     public ContentSinkOutputStream(Content.Sink sink)
     {
@@ -55,7 +56,7 @@ public class ContentSinkOutputStream extends OutputStream
         }
         catch (Throwable x)
         {
-            throw IO.rethrow(x);
+            handleException(x);
         }
     }
 
@@ -69,7 +70,7 @@ public class ContentSinkOutputStream extends OutputStream
         }
         catch (Throwable x)
         {
-            throw IO.rethrow(x);
+            handleException(x);
         }
     }
 
@@ -83,12 +84,20 @@ public class ContentSinkOutputStream extends OutputStream
         }
         catch (Throwable x)
         {
-            throw IO.rethrow(x);
+            handleException(x);
         }
     }
 
     public void close(Callback callback) throws IOException
     {
         sink.write(true, null, callback);
+    }
+
+    private void handleException(Throwable x) throws IOException
+    {
+        if (failed)
+            throw new IOException(x.toString());
+        failed = true;
+        throw IO.rethrow(x);
     }
 }

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/content/ContentSinkOutputStreamTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/content/ContentSinkOutputStreamTest.java
@@ -1,0 +1,58 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io.content;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.EofException;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.not;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class ContentSinkOutputStreamTest
+{
+    @Test
+    public void testNoDuplicateExceptionThrown()
+    {
+        EofException eofException = new EofException();
+        Content.Sink sink = (last, byteBuffer, callback) -> callback.failed(eofException);
+
+        StringWriter stringWriter = new StringWriter();
+        try (OutputStream body = new ContentSinkOutputStream(sink))
+        {
+            try
+            {
+                body.write('a');
+                fail("expected IOException in write()");
+            }
+            catch (IOException e)
+            {
+                body.flush();
+                fail("expected IOException in flush()");
+            }
+        }
+        catch (IOException ex)
+        {
+            ex.printStackTrace(new PrintWriter(stringWriter));
+        }
+        assertThat(stringWriter.toString(), not(containsString("CIRCULAR REFERENCE")));
+    }
+}

